src/foam/nanos/auth/twofactor/TwoFactorAuthService.js

```diff
@@ -32,7 +33,7 @@ foam.CLASS({
       name: 'check',
       javaCode: `
         Session session = x.get(Session.class);
-        User user = (User) x.get("user");
+        User user = ((Subject) x.get("subject")).getUser();
```

Comment:
Interesting, even before I thought this would have been the real user.

Reply:
x.get("user") is the real user only in case when agent is null. If user is acting as somebody else then the real user is stored in "agent", the acting user is stored in "user". So currently whenever code does x.get("user") it only cares for acting user. There are places with code like `agent != null ? agent : user` or just `x.get("agent")` -  thats when people want to get the real user(the logged in one). 
`subject.getUser()` - returns the last acting user; if user never acted as somebody else, returns logged in user.
`subject.getRealUser()` - always returns logged in user